conf.c

```diff
@@ -93,168 +128,482 @@ chomp(char *str)
+	initialize_configuration_setting(CONF_SPOOLDIR, NULL, true, DEFAULT_SPOOLDIR);
+	initialize_configuration_setting(CONF_AUTHPATH, NULL, true, NULL);
+	initialize_configuration_setting(CONF_CERTFILE, NULL, true, NULL);
+	initialize_configuration_setting(CONF_MAILNAME, NULL, true,  NULL);
+	initialize_configuration_setting(CONF_MASQUERADE, NULL, true, NULL);
+	initialize_configuration_setting(CONF_STARTTLS, NULL, false, NULL);
+	initialize_configuration_setting(CONF_FINGERPRINT, &check_fingerprint_configuration, true, NULL);
+	initialize_configuration_setting(CONF_TLS_OPP, NULL, false, NULL);
+	initialize_configuration_setting(CONF_SECURETRANSFER, NULL, false, NULL);
+	initialize_configuration_setting(CONF_DEFER, NULL, false, NULL);
+	initialize_configuration_setting(CONF_INSECURE, NULL, false, NULL);
+	initialize_configuration_setting(CONF_FULLBOUNCE, NULL, false, NULL);
+	initialize_configuration_setting(CONF_NULLCLIENT, NULL, false, NULL);
+}
+
+static void initialize_configuration_setting(
+		const char* identifier,
+		check_ptr_t func_ptr, bool needs_value,
+		const char* default_value
+		)
+{
+	struct config_item_t *item;
+
+	if(identifier == NULL)
+		return;
 
-	while (!feof(conf)) {
-		if (fgets(line, sizeof(line), conf) == NULL)
+	/* Check if the setting has already been initialized */
+	SLIST_FOREACH(item, &config_head, next_item)
+		if(!strcmp(identifier, item->identifier))
+			/* Already in list */
+			errlogx(EX_SOFTWARE, "Trying to initialize setting '%s' more than once", identifier);
+
+	item = calloc(1, sizeof(struct config_item_t));
+	if(item == NULL)
+		errlog(EX_OSERR,"Error allocating memory\n");
+		/* exit */
+
+	/* We use strdup() here so that we'll be able to properly clean up everything */
+	item->identifier = strdup(identifier);
+	item->checkFunction = func_ptr;
+	item->needs_value = needs_value;
+
+	if(default_value != NULL)
+		item->str_value = strdup(default_value);
+	else
+		item->str_value = NULL;
+
+	SLIST_INSERT_HEAD(&config_head, item, next_item);
+}
+
+int try_to_set_configuration_setting(char *identifier, char *value)
+{
+	struct config_item_t *item = NULL;
+	bool found = false;
+
+	SLIST_FOREACH(item, &config_head, next_item)
+		if(!strcmp(identifier, item->identifier))
+		{
+			found = true;
 			break;
-		lineno++;
+		}
 
-		chomp(line);
+	if(found)		/* item is pointing to the correct item now */
+	{
+		 /* Let's see if it requires a value */
+		if(item->needs_value)
+		{
+			/* Check values */
+			if(value == NULL)
+				/* Item requires a value, but no value has been provided! */
+				return EX_CONFIG;
 
-		/* We hit a comment */
-		if (strchr(line, '#'))
-			*strchr(line, '#') = 0;
+			/* Call the check function, if it exists */
+			if(item->checkFunction != NULL)
+				if(item->checkFunction(identifier, value) != 0)
+					return EX_CONFIG;
 
-		data = line;
-		word = strsep(&data, EQS);
+			/* Else let's store the value and return */
+			if(item->str_value != NULL)
```

Comment:
you can free(NULL), so no need to test.

Reply:
Well, yes, but in this case, it's more likely for item->str_value to be NULL than not. It can only be non-NULL if there's a duplicate setting in the config-file or there's a default-value (3 settings with vs 6 without a default value). So, we'd probably save a function call in most cases at the expense of an extra test in less cases.

Reply:
I don't understand why the likelihood of being NULL matters.  removing the conditional simplifies the code, that's all.

Reply:
Never mind, I'll just remove the check before the free.